blogs/2020/02/14/optimizing-ci.md

```diff
@@ -0,0 +1,117 @@
+```
+
+Here, an alias of `Packages` is appended to the environment variable output, allowing us to cache NPM packages and compilation output in a single build job. We were finally deduplicating a lot of the CI work that can now be executed only once and shared across platform-specific agents.
+
+There was still room for one final optimization, given a specific use case: build resubmissions. We sometimes must retrigger builds on previously built commits since tests might be flaky or some agents might randomly fail. Ideally, the shared agent would not restore or recompile the common code but simply defer to the platform-dependent agents to perform their work. The issue we noticed was that the compilation cache packages were massive and restoring them would take around 8 minutes—all for nothing since the shared agent would simply yield control if that cache existed. We addressed this issue by implementing a `dryRun` feature to allow us to check for a cache package’s existence without restoring it—effectively cutting 8 minutes off our build resubmissions. Using the `dryRun` parameter looks like this:
+
+```yml
+- task: 1ESLighthouseEng.PipelineArtifactCaching.RestoreCacheV1.RestoreCache@1
+  inputs:
+    keyfile: commit
+    targetfolder: output
+    vstsFeed: "$(ArtifactFeed)"
+    dryRun: true
+
+- script: |
+    npm run compile install
+  displayName: Install Dependencies
+  condition: ne(variables['CacheExists'], 'true')
+```
+
+Notice this also introduced a new `CacheExists` variable, which works together with the `dryRun` parameter.
+
+## Results
+
+After all these changes were implemented, we saw drastic reductions in total build time. The following table shows the change in total build times for each platform that VS Code targets:
+
+| Platform     | Before | After | Time Savings |
+|--------------|--------|-------|--------------|
+| Windows      | 58s     | 44s    |  24%         |
+| Windows 32   | 59s     | 46s    | 22%          |
+| Linux        | 38s     | 23s    | 39%          |
+| macOS        | 68s     | 42s    | 38%          |
+| Linux ARM    | 22s     | 21s    | 5%           |
+| Linux Alpine | 23s     | 26s    | -13%         |
+
+Here's a pretty chart, to get a visual perception of the improvements:
+
+![chart](chart.png)
+
+Build resubmissions saw a drastic improvement, since the shared agent tasks can be skipped altogether. Here are some numbers for macOS, for example:
+
+| Platform     | Before | After | Time Savings |
+|--------------|--------|-------|--------------|
+| macOS        | 68s     | 34s    | 50%          |
+
+In total, we were thrilled to see a combined ~50% reduction in VS Code’s CI build times! The best news is that you can draw inspiration from [our build definitions](https://github.com/microsoft/vscode/tree/master/build/azure-pipelines) to realize build time improvements of your own.
+
+Happy Caching,
+
+Ethan Dennis, Azure DevOps Senior Software Engineer
```

Comment:
My product group should be listed as `Developer Services`

Reply:
```suggestion
Ethan Dennis, Developer Services Senior Software Engineer
```